File: ip_calculator.js
// ip_calculator.js — JavaScript версия

const fs = require('fs');
const os = require('os');

class IPv6Calculator {
    constructor(cidr) {
        this.cidr = cidr;
        this.parse();
    }

    parse() {
        const parts = this.cidr.split('/');
        if (parts.length !== 2) throw new Error('Неверный формат CIDR');
        this.address = parts[0];
        this.prefix = parseInt(parts[1]);
        if (this.prefix < 0 || this.prefix > 128) throw new Error('Префикс должен быть 0-128');
        // Неполная реализация, используем простые вычисления
        // В реальности нужно использовать библиотеку, но для демонстрации упростим
        this.network = this.address; // Упрощённо
        this.mask = this.calculateMask();
        this.numAddresses = BigInt(1) << BigInt(128 - this.prefix);
        this.firstHost = this.address; // упрощённо
        this.lastHost = this.address; // упрощённо
    }

    calculateMask() {
        // Возвращает строку маски
        return `/${this.prefix}`;
    }

    getInfo() {
        return {
            ip: this.address,
            mask: this.mask,
            cidr: `/${this.prefix}`,
            network: this.network,
            first_host: this.firstHost,
            last_host: this.lastHost,
            num_addresses: this.numAddresses.toString(),
            version: 'IPv6',
            prefixlen: this.prefix
        };
    }

    printInfo() {
        const info = this.getInfo();
        console.log('\x1b[36m🌐 IP Calculator (IPv6) (JavaScript)\x1b[0m');
        console.log(`Входные данные: ${this.cidr}`);
        console.log();
        console.log('\x1b[32m─────────────────────────────────────────\x1b[0m');
        console.log(`IP-адрес:       ${info.ip}`);
        console.log(`Маска (CIDR):   ${info.cidr} (${info.mask})`);
        console.log(`Сетевой адрес:  ${info.network}`);
        console.log(`Количество адресов: ${info.num_addresses}`);
        console.log(`Первый хост:    ${info.first_host}`);
        console.log(`Последний хост: ${info.last_host}`);
        console.log(`Версия:         ${info.version}`);
        console.log('\x1b[32m─────────────────────────────────────────\x1b[0m');
    }

    saveJSON(filename = 'ip_calc_output.json') {
        const data = {
            timestamp: new Date().toISOString(),
            input: this.cidr,
            result: this.getInfo()
        };
        fs.writeFileSync(filename, JSON.stringify(data, null, 2));
        console.log(`\x1b[32m💾 Сохранено JSON: ${filename}\x1b[0m`);
    }

    saveCSV(filename = 'ip_calc_output.csv') {
        const info = this.getInfo();
        let csv = 'Parameter,Value\n';
        for (const [key, val] of Object.entries(info)) {
            csv += `${key},${val}\n`;
        }
        fs.writeFileSync(filename, csv);
        console.log(`\x1b[32m💾 Сохранено CSV: ${filename}\x1b[0m`);
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log('Usage: node ip_calculator.js <IPv6/CIDR>');
        console.log('Пример: node ip_calculator.js 2001:db8::/32');
        process.exit(1);
    }
    const cidr = args[0];
    try {
        const calc = new IPv6Calculator(cidr);
        calc.printInfo();
        calc.saveJSON();
        calc.saveCSV();
    } catch (err) {
        console.error(`\x1b[31m❌ Ошибка: ${err.message}\x1b[0m`);
        process.exit(1);
    }
}

if (require.main === module) main();

File: ip_calculator.rs
// ip_calculator.rs — Rust версия

use std::env;
use std::net::{Ipv6Addr, Ipv6Network};
use std::fs;
use std::collections::HashMap;
use serde::{Deserialize, Serialize};
use serde_json;

#[derive(Debug, Serialize, Deserialize)]
struct IPv6Info {
    ip: String,
    mask: String,
    cidr: String,
    network: String,
    first_host: String,
    last_host: String,
    num_addresses: String,
    version: String,
    prefixlen: u8,
}

fn calculate_ipv6(cidr: &str) -> Result<IPv6Info, Box<dyn std::error::Error>> {
    let network: Ipv6Network = cidr.parse()?;
    let ip = network.network();
    let mask = network.mask();
    let prefix = network.prefix();
    let num = 1u128.checked_shl(128 - prefix as u32).unwrap_or(0);
    let num_str = if num == 0 { "0".to_string() } else { num.to_string() };

    // Первый и последний (упрощённо)
    let first_host = ip.to_string();
    let last_host = ip.to_string();

    Ok(IPv6Info {
        ip: ip.to_string(),
        mask: mask.to_string(),
        cidr: format!("/{}", prefix),
        network: ip.to_string(),
        first_host,
        last_host,
        num_addresses: num_str,
        version: "IPv6".to_string(),
        prefixlen: prefix,
    })
}

fn print_info(info: &IPv6Info, input: &str) {
    println!("\x1b[36m🌐 IP Calculator (IPv6) (Rust)\x1b[0m");
    println!("Входные данные: {}", input);
    println!();
    println!("\x1b[32m─────────────────────────────────────────\x1b[0m");
    println!("IP-адрес:       {}", info.ip);
    println!("Маска (CIDR):   {} ({})", info.cidr, info.mask);
    println!("Сетевой адрес:  {}", info.network);
    println!("Количество адресов: {}", info.num_addresses);
    println!("Первый хост:    {}", info.first_host);
    println!("Последний хост: {}", info.last_host);
    println!("Версия:         {}", info.version);
    println!("\x1b[32m─────────────────────────────────────────\x1b[0m");
}

fn save_json(info: &IPv6Info, input: &str) -> Result<(), Box<dyn std::error::Error>> {
    let data = serde_json::json!({
        "timestamp": chrono::Utc::now().to_rfc3339(),
        "input": input,
        "result": info
    });
    let json = serde_json::to_string_pretty(&data)?;
    fs::write("ip_calc_output.json", json)?;
    println!("\x1b[32m💾 Сохранено JSON: ip_calc_output.json\x1b[0m");
    Ok(())
}

fn save_csv(info: &IPv6Info) -> Result<(), Box<dyn std::error::Error>> {
    let mut csv = String::from("Parameter,Value\n");
    csv.push_str(&format!("ip,{}\n", info.ip));
    csv.push_str(&format!("mask,{}\n", info.mask));
    csv.push_str(&format!("cidr,{}\n", info.cidr));
    csv.push_str(&format!("network,{}\n", info.network));
    csv.push_str(&format!("first_host,{}\n", info.first_host));
    csv.push_str(&format!("last_host,{}\n", info.last_host));
    csv.push_str(&format!("num_addresses,{}\n", info.num_addresses));
    csv.push_str(&format!("version,{}\n", info.version));
    csv.push_str(&format!("prefixlen,{}\n", info.prefixlen));
    fs::write("ip_calc_output.csv", csv)?;
    println!("\x1b[32m💾 Сохранено CSV: ip_calc_output.csv\x1b[0m");
    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        println!("Usage: cargo run -- <IPv6/CIDR>");
        println!("Пример: cargo run -- 2001:db8::/32");
        std::process::exit(1);
    }
    let input = &args[1];
    let info = calculate_ipv6(input)?;
    print_info(&info, input);
    save_json(&info, input)?;
    save_csv(&info)?;
    Ok(())
}
